editor/editor_properties.cpp

```diff
@@ -2182,8 +2182,97 @@ void EditorPropertyResource::_menu_option(int p_which) {
 				EditorNode::get_singleton()->get_scene_tree_dock()->open_script_dialog(Object::cast_to<Node>(get_edited_object()));
+			} else {
+
+				// Based upon editor/scene_tree_dock.cpp#_tool_selected (case TOOL_ATTACH_SCRIPT); edited
+				// to account for exported script properties, resources, and sub-resources
+
+				Ref<Script> existing = get_edited_object()->get(get_edited_property());
+
+				String inherits;
+				if (existing.is_valid()) {
+					// A script already exists in this property, so inherit from that script
+					for (int i = 0; i < ScriptServer::get_language_count(); i++) {
+						ScriptLanguage *l = ScriptServer::get_language(i);
+						if (l->get_type() == existing->get_class()) {
+							String name = l->get_global_class_name(existing->get_path());
+							if (ScriptServer::is_global_class(name) && EDITOR_GET("interface/editors/derive_script_globals_by_name").operator bool()) {
+								inherits = name;
+							} else if (l->can_inherit_from_file()) {
+								inherits = "\"" + existing->get_path() + "\"";
+							}
+							break;
+						}
+					}
+				} else if (get_edited_property() == "script") {
+					// This is the script property of an object's sub-property, so inherit from that object's class
+					inherits = get_edited_object()->get_class();
+				} else {
+					inherits = "Reference";
+				}
+
+				String path;
+				// Path creation:
+				// Get the object's basename
+				//	If the object has no basename, get the scene root's filename
+				//		If the scene root has a filename, append the object's name/class to it
+				//		If it does not, append the object's name/class to "res://"
+				// If the property we're editing is not "script", append the property name to the filename
+				if (Object::cast_to<Node>(get_edited_object()) &&
+						EditorNode::get_singleton()->get_editor_data().get_edited_scene_root()->is_a_parent_of(Object::cast_to<Node>(get_edited_object()))) {
+					// This is a node in the scene tree, but it is not the script property
+					Node *edited_node = Object::cast_to<Node>(get_edited_object());
+
+					path = edited_node->get_filename().get_basename();
+					if (path == "") {
+						String root_path = EditorNode::get_singleton()->get_editor_data().get_edited_scene_root()->get_filename();
+						if (root_path == "") {
+							path = String("res://").plus_file(edited_node->get_name());
+						} else {
+							path = root_path.get_base_dir().plus_file(edited_node->get_name());
+						}
+					}
+				} else if (Object::cast_to<Resource>(get_edited_object())) {
+					RES edited_res = Object::cast_to<Resource>(get_edited_object());
+
+					path = edited_res->get_path().get_basename();
+					if (path == "") {
+						String root_path = EditorNode::get_singleton()->get_editor_data().get_edited_scene_root()->get_filename();
+						if (root_path == "") {
+							path = String("res://").plus_file(edited_res->get_name());
+						} else {
+							path = root_path.get_base_dir().plus_file(edited_res->get_name());
+						}
```

Comment:
Would it be more appropriate to put this section of code (as well as the copies for Node and the else statement) in a macro, for re-use in all three cases? I would normally just go ahead and do so, but I'm not sure if such a move would be appreciated (considering the only macros I ever see in the engine are the error macros), or if there is perhaps a better way of doing that.